### PR TITLE
Active learning abstraction

### PR DESCRIPTION
**Pull request by laserkelvin**
# PhysicsNeMo Pull Request

## Description

This PR introduces the initial components that make up the active learning abstraction within `physicsnemo`:

- Adds a new namespace, `physicsnemo.active_learning`, with multiple submodules and a `README.md` that briefly explains the module layout until proper documentation is added. As an overview:
    - `protocols` defines interfaces (abstract classes) for the various components used by an active learning workflow
    - `driver` defines a concrete orchestrator that "drives" the end-to-end active learning process in an automated fashion with the ability to restart from checkpoints
    - `config` defines `dataclass`es that controls the flow of active learning
- Adds accompanying unit tests under `test/active_learning`
- Adds a simple example case that demonstrates active learning within `physicsnemo` with a well-known "2 moons" dataset: users can parse this example to understand what components need to be defined, and how they are composed together for `Driver` to execute.

## Checklist

- [x] I am familiar with the [Contributing Guidelines](https://github.com/NVIDIA/physicsnemo/blob/main/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.
- [x] The [CHANGELOG.md](https://github.com/NVIDIA/physicsnemo/blob/main/CHANGELOG.md) is up to date with these changes.
- [ ] An [issue](https://github.com/NVIDIA/physicsnemo/issues) is linked to this pull request.

## Dependencies

No new dependencies were added.

### CODE REVIEW

File: physicsnemo/active_learning/config.py [hunk 1]
```diff
@@ -0,0 +1,808 @@
...
+    scheduler_cls: type[_LRScheduler] | None = None
+    scheduler_kwargs: dict[str, Any] = field(default_factory=dict)
+
+    def __post_init__(self) -> None:
+        """Validate optimizer configuration."""
+        # Validate learning rate if present
+        if "lr" in self.optimizer_kwargs:
+            lr = self.optimizer_kwargs["lr"]
+            if not isinstance(lr, (int, float)) or lr <= 0:
+                raise ValueError(f"Learning rate must be positive, got {lr}")
```
[the remark this answers is not in the document]
Comment (laserkelvin) [reply physicsnemo/active_learning/config.py:80]: I'm leaving this out, as a user would have to go out of their way to trigger this edge case

File: physicsnemo/active_learning/config.py [hunk 2]
```diff
@@ -0,0 +1,808 @@
...
+        if self.val_datapool is not None and not hasattr(self.val_datapool, "__len__"):
+            raise ValueError("val_datapool must implement __len__")
+
+        # Validate training loop is callable
+        if not callable(self.train_loop_fn):
+            raise ValueError("train_loop_fn must be callable")
+
+        # set the same value for fine tuning epochs if not provided
+        if not self.max_fine_tuning_epochs:
+            self.max_fine_tuning_epochs = self.max_training_epochs
```
[the remark this answers is not in the document]
Comment (laserkelvin) [reply physicsnemo/active_learning/config.py]: Fixed in [cb122cb](https://github.com/NVIDIA/physicsnemo/pull/1174/commits/cb122cb400c6a57576814c2c84de768913b2a0d3)

File: physicsnemo/active_learning/loop.py [hunk 3]
```diff
@@ -0,0 +1,528 @@
...
+                ) from e
+        if not validate_step_fn:
+            validate_step_fn = self.validate_step_fn
+        if validate_step_fn:
+            val_func_id = id(validate_step_fn)
+            try:
+                validate_step_fn = StaticCaptureEvaluateNoGrad(
+                    model=model, amp_type=self.amp_type, **self.capture_kwargs
+                )(validate_step_fn)
+                self.capture_functions[val_func_id] = validate_step_fn
```
[the remark this answers is not in the document]
Comment (laserkelvin) [reply physicsnemo/active_learning/loop.py]: Fixed the caching mechanism in general in [2717bc8](https://github.com/NVIDIA/physicsnemo/pull/1174/commits/2717bc81dae3780cfcfc1cd6dad3d128a22ecc6d)

File: examples/active_learning/moons/moon_strategies.py [hunk 4]
```diff
@@ -0,0 +1,235 @@
...
+
+from physicsnemo.active_learning import registry
+from physicsnemo.active_learning.protocols import (
+    DriverProtocol,
+    LabelStrategy,
+    MetrologyStrategy,
+    QueryStrategy,
+)
+
+__all__ = ["ClassifierUQQuery", "DummyLabelStrategy"]
```
[the remark this answers is not in the document]
Comment (laserkelvin) [reply examples/active_learning/moons/moon_strategies.py]: Done in 0c4aab4fce57d2e682a95b87788afd3c95a4881c

File: physicsnemo/active_learning/logger.py [hunk 5]
```diff
@@ -0,0 +1,327 @@
...
+    console_handler.setLevel(logging.DEBUG)
+    console_handler.setFormatter(
+        ContextFormatter("%(asctime)s - %(name)s - %(levelname)s - %(message)s")
+    )
+    console_handler.addFilter(ContextInjectingFilter())
+    logger.addHandler(console_handler)
+
+    # Set up file handler with JSON formatting
+    log_file = log_dir_path / f"{run_id}.log"
+    file_handler = logging.FileHandler(log_file, mode="w")
```
[the remark this answers is not in the document]
Comment (laserkelvin) [reply physicsnemo/active_learning/logger.py:280]: Run ids are unique so they won't overwrite, anyway.

File: physicsnemo/active_learning/_registry.py [hunk 6]
```diff
@@ -0,0 +1,330 @@
...
+        --------
+        >>> strategy = registry.construct("my_strategy", param1=42, param2="test")
+        """
+        # Check if the class is registered
+        if cls_name not in self._registry:
+            available = ", ".join(self._registry.keys()) if self._registry else "none"
+            raise KeyError(
+                f"Protocol '{cls_name}' is not registered. "
+                f"Available protocols: {available}"
+            )
```
[the remark this answers is not in the document]
Comment (laserkelvin) [reply physicsnemo/active_learning/_registry.py]: Fixed in 5f31a6d9cc12a32490085b0fd8d606fa72914b5e

File: examples/active_learning/moons/moon_data.py [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+    outer_y = torch.sin(outer_grid)
+    inner_x = 1 - torch.cos(inner_grid)
+    inner_y = 1 - torch.sin(inner_grid) - 0.5
+    outer = torch.stack([outer_x, outer_y], dim=-1)
+    inner = torch.stack([inner_x, inner_y], dim=-1)
+    X_values = torch.cat([outer, inner], dim=0)
+    # add some noise to the coordinates
+    X_values += torch.randn_like(X_values) * sigma
+    y_values = torch.zeros(n_samples)
+    y_values[inner_grid.shape[0] :] = 1
```
[the remark this answers is not in the document]
Comment (laserkelvin) [reply examples/active_learning/moons/moon_data.py]: Fixed in a31cd18beb7db85156b20f7586ef7f6c708bb7f9

File: physicsnemo/active_learning/loop.py [hunk 8]
```diff
@@ -0,0 +1,528 @@
...
+                """
+            )
+        if not device and not self.device:
+            device = torch.get_default_device()
+        if not dtype and not self.dtype:
+            dtype = torch.get_default_dtype()
+        # if a device is specified, move the model
+        if device and device != model.device:
+            # not 100% sure this will trigger issues with the optimizer
+            model = model.to(self.device)
```
[the remark this answers is not in the document]
Comment (laserkelvin) [reply physicsnemo/active_learning/loop.py]: Fixed in 8b8d3246138411cf2605aaf377e2da944a685fb9

File: physicsnemo/active_learning/driver.py [hunk 9]
```diff
@@ -0,0 +1,1430 @@
...
+                    f"Invalid LR scheduler kwargs for {opt_cfg.scheduler_cls}; {e}"
+                )
+            self.lr_scheduler = opt_cfg.scheduler_cls(
+                self.optimizer, **opt_cfg.scheduler_kwargs
+            )
+        else:
+            self.lr_scheduler = None
+        # in the case where we want to reset optimizer states between active learning steps
+        if self.config.reset_optim_states:
+            self._original_optim_state = deepcopy(self.optimizer.state_dict())
```
[the remark this answers is not in the document]
Comment (laserkelvin) [reply physicsnemo/active_learning/driver.py]: Fixed in ad1c90c36f56f7e7246e5d446acd9419d9061c93

File: physicsnemo/active_learning/driver.py [hunk 10]
```diff
@@ -0,0 +1,1430 @@
...
+        else:
+            if self.config.device is not None:
+                self.learner = self.learner.to(self.config.device, self.config.dtype)
+        # assume all device management is done via the dist_manager, so at this
+        # point the model is on the correct device and we can set up the optimizer
+        # if we intend to train
+        if not self.config.skip_training and not self.is_optimizer_configured:
+            self.configure_optimizer()
+        if not self.config.skip_training and self.config.reset_optim_states:
+            self.optimizer.load_state_dict(self._original_optim_state)
```
[the remark this answers is not in the document]
Comment (laserkelvin) [reply physicsnemo/active_learning/driver.py]: Fixed in 1057ea367c52bd254e83f1b76122e2ef0f5a5215

File: examples/active_learning/moons/moon_strategies.py [hunk 11]
```diff
@@ -0,0 +1,235 @@
...
+        Returns
+        -------
+        float
+            Precision score for class 0.
+        """
+        true_positives = (true_labels == 0).sum().item()
+        predicted_positives = (pred_labels == 0).sum().item()
+        if predicted_positives == 0:
+            return 0.0
+        return true_positives / (predicted_positives + true_positives)
```
[the remark this answers is not in the document]
Comment (laserkelvin) [reply examples/active_learning/moons/moon_strategies.py]: For reference, `true_positives / predicted_positives` can be greater than 1.

I'm also referring to the `sklearn` docs: https://scikit-learn.org/stable/auto_examples/model_selection/plot_precision_recall.html

[the remark this answers is not in the document]
Comment (laserkelvin) [reply examples/active_learning/moons/moon_strategies.py]: Turns out I was wrong and overlooked the "true" positives as the intersection.

I fixed this, and I also flipped the "positive" as label 1 in dfc40e0d9315e9d6987a76159e1e4daf67317bec

**Comment by coreyjadams** [comment]
If you use `torch.compile(mode="max-autotune")` or `mode=reduce-latency`, does that give the same benefits as static capture?  Alternatively, you can use `torch.cuda.make_graphed_callable` to generate the cuda graph.

If these upstream tools are delivering the functionality just as well, I think we should use them instead.

**Comment by laserkelvin** [comment]
> Member

I agree with your point - but to your question, should we update static capture classes so that users can pass `kwargs` to change how `torch.compile` operates on the model?

That'll give it the flexibility needed for users to tinker with getting architecture-specific performance gains out of `torch.compile`

File: physicsnemo/active_learning/loop.py [hunk 12]
```diff
@@ -0,0 +1,534 @@
...
+                    )
+                    model.zero_grad(set_to_none=True)
+                    loss = train_step_fn(model, batch, *args, **kwargs)
+                    log.log_minibatch({"train_loss": loss.detach().item()})
+                    # normally, static capture will call backward because of AMP
+                    if not self.enable_static_capture:
+                        loss.backward()
+                    optimizer.step()
+                    if lr_scheduler:
+                        lr_scheduler.step()
```
[the remark this answers is not in the document]
Comment (laserkelvin) [reply physicsnemo/active_learning/loop.py:505]: This is a valid point - I'll create an issue to track

File: examples/active_learning/moons/moon_example.py [hunk 13]
```diff
@@ -0,0 +1,190 @@
...
+    # just some sanity checks
+    if not (
+        len(dataset.train_indices)
+        == int(dataset.initial_samples * dataset.total_samples)
+        + driver_config.max_active_learning_steps
+        * strategy_config.query_strategies[0].max_samples
+    ):
+        raise RuntimeError(
+            "Number of samples added to the training pool inconsistent with expected value."
+        )
```
[the remark this answers is not in the document]
Comment (laserkelvin) [reply examples/active_learning/moons/moon_example.py:183]: This is not relevant for this example, as there is no early stopping functionality.

File: physicsnemo/active_learning/loop.py [hunk 14]
```diff
@@ -0,0 +1,534 @@
...
+        """
+        # Load model weights separately
+        if isinstance(model, Module):
+            model_path = checkpoint_dir / "model.mdlus"
+            if model_path.exists():
+                model.load(str(model_path))
+        else:
+            model_state_path = checkpoint_dir / "model_state.pt"
+            if model_state_path.exists():
+                state_dict = torch.load(model_state_path, map_location="cpu")
```
[the remark this answers is not in the document]
Comment (laserkelvin) [reply physicsnemo/active_learning/loop.py:293]: Good bot
